File: 0015 三数之和/main.py
from typing import List
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        tmpSet = []
        nums.sort()
        lasti = None
        lastj = None
        lastk= None
        for i in range(len(nums)-2):

            valuei = nums[i]
            if(lasti==nums[i]):
                continue

            li = i+1
            ri = len(nums)-1                

            lastj = None
            lastk= None
            while(li != ri):
                # print(nums, valuei,nums[li], nums[ri])
                j=nums[li]
                k=nums[ri]
                if(lastj==j):
                    li+=1
                    continue
                if(lastk==k):
                    ri-=1
                    continue
                if(0 == nums[li]+nums[ri]+valuei):
                    tmpSet.append([valuei, nums[li], nums[ri]])
                    lastk=k
                    ri -= 1
                elif(0 <= nums[li]+nums[ri]+valuei):
                    ri -= 1
                    lastk=k
                else:
                    li += 1
                    lastj=j
            lasti=nums[i]
        return tmpSet
a=[3,0,-2,-1,1,2]
b=Solution()
print(b.threeSum(a))

File: 0015 三数之和/main.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        vector<vector<int>> res;
        int target=0;
        for(int i=0;i<nums.size();i++){
            if(i!=0 && nums[i] ==nums[i-1])
                continue;
            target=-nums[i];
            int li=i+1;
            int ri=nums.size()-1;
            while(ri>li){
                if(target > nums[li]+nums[ri] || (li > i+1 && nums[li]==nums[li-1])){
                    li++;
                }else if(target < nums[li] + nums[ri] || ( ri < nums.size()-1 && nums[ri] == nums[ri+1] )){
                    ri--;
                }else{
                    vector<int> tmp={ nums[i] , nums[li], nums[ri] };
                    res.push_back(tmp);
                    li++;
                }
            }
        }
        return res;
    }
};